From 0dabff312fd2b531186ba00444c810f99177685f Mon Sep 17 00:00:00 2001
From: David Gibson <dgibson@redhat.com>
Date: Mon, 15 Sep 2014 07:12:34 -0400
Subject: [virt] kvm/ppc: book3s/hv - Fix machine check delivery to guest

Message-id: <1410765214-16377-19-git-send-email-dgibson@redhat.com>
Patchwork-id: 94711
O-Subject: [PATCH 18/78] KVM: PPC: Book3S HV: Fix machine check delivery to guest
Bugzilla: 1123145 1123133 1123367
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1123145

Signed-off-by: David Gibson <dgibson@redhat.com>

commit 000a25ddb7106cdcb34e7f6c7547e5b2354b6645
Author: Paul Mackerras <paulus@samba.org>
Date:   Mon May 26 19:48:41 2014 +1000

    KVM: PPC: Book3S HV: Fix machine check delivery to guest

    The code that delivered a machine check to the guest after handling
    it in real mode failed to load up r11 before calling kvmppc_msr_interrupt,
    which needs the old MSR value in r11 so it can see the transactional
    state there.  This adds the missing load.

    Signed-off-by: Paul Mackerras <paulus@samba.org>
    Signed-off-by: Alexander Graf <agraf@suse.de>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kvm/book3s_hv_rmhandlers.S b/arch/powerpc/kvm/book3s_hv_rmhandlers.S
index 8069a9f..a1f1b1e 100644
--- a/arch/powerpc/kvm/book3s_hv_rmhandlers.S
+++ b/arch/powerpc/kvm/book3s_hv_rmhandlers.S
@@ -2252,6 +2252,7 @@ machine_check_realmode:
  bne 2f /* Continue guest execution. */
  /* If not, deliver a machine check.  SRR0/1 are already set */
  li r10, BOOK3S_INTERRUPT_MACHINE_CHECK
+ ld r11, VCPU_MSR(r9)
  bl kvmppc_msr_interrupt
 2: b fast_interrupt_c_return
 
-- 
1.7.1